connector/src/main/java/com/google/cloud/spark/bigquery/AvroSchemaConverter.java

```diff
@@ -0,0 +1,287 @@
+            .addToSchema(builder.bytesType());
+      } else {
+        throw new IllegalArgumentException(
+            "Decimal type is too wide to fit in BigQuery Numeric format");
+      }
+    }
+    if (dataType instanceof StringType) {
+      return builder.stringType();
+    }
+    if (dataType instanceof TimestampType) {
+      // return builder.TIMESTAMP; FIXME: Restore this correct conversion when the Vortex
+      // team adds microsecond support to their backend
+      return LogicalTypes.timestampMicros().addToSchema(builder.longType());
+    }
+    if (dataType instanceof DateType) {
+      return LogicalTypes.date().addToSchema(builder.intType());
+    }
+    if (dataType instanceof ArrayType) {
+      return builder
+          .array()
+          .items(
+              sparkTypeToRawAvroType(
+                  ((ArrayType) dataType).elementType(),
+                  ((ArrayType) dataType).containsNull(),
+                  recordName));
+    }
+    if (dataType instanceof StructType) {
+      SchemaBuilder.FieldAssembler<Schema> fieldsAssembler = builder.record(recordName).fields();
+      for (StructField field : ((StructType) dataType).fields()) {
+        Schema avroType = sparkTypeToRawAvroType(field.dataType(), field.nullable(), field.name());
+        fieldsAssembler.name(field.name()).type(avroType).noDefault();
+      }
+      return fieldsAssembler.endRecord();
+    }
+    if (dataType instanceof UserDefinedType) {
+      DataType userDefinedType = ((UserDefinedType) dataType).sqlType();
+      return sparkTypeToRawAvroType(userDefinedType, recordName, builder);
+    }
+    if (dataType instanceof MapType) {
+      throw new IllegalArgumentException(SchemaConverters.MAPTYPE_ERROR_MESSAGE);
+    } else {
+      throw new IllegalArgumentException("Data type not expected: " + dataType.simpleString());
+    }
+  }
+
+  public static GenericData.Record sparkRowToAvroGenericData(
+      InternalRow row, StructType sparkSchema, Schema avroSchema) {
+    StructConverter structConverter = new StructConverter(sparkSchema, avroSchema);
+    return structConverter.convert(row);
+  }
+
+  static Schema resolveNullableType(Schema avroType, boolean nullable) {
+    if (nullable && avroType.getType() != Schema.Type.NULL) {
+      // avro uses union to represent nullable type.
+      List<Schema> fields = avroType.getTypes();
+      Preconditions.checkArgument(
+          fields.size() == 2, "Avro nullable filed should be represented by a union of size 2");
+      Optional<Schema> actualType =
+          fields.stream().filter(field -> field.getType() != Schema.Type.NULL).findFirst();
+      return actualType.orElseThrow(
+          () -> new IllegalArgumentException("No actual type has been found in " + avroType));
+    } else {
+      return avroType;
+    }
+  }
+
+  static Converter createConverterFor(DataType sparkType, Schema avroType) {
+    if (sparkType instanceof NullType && avroType.getType() == Schema.Type.NULL) {
```

Comment:
Please feel free to ignore this, but there is a bunch of type checks and casts in the code. This is where Scala pattern matching shines, but since this is Java, I was wondering if using a visitor pattern would make this cleaner.

Reply:
I agree the code is cumbersome, but as it is built on Scala classes, meant to be used with pattern matching, they do not support the visitor pattern. Also specifically here, there's a dual matching - both to spark (catalyst) type and the avro type